File: java/src/test/java/com/corntest/RandomMathOperationsTest.java
package com.corntest;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Test class for RandomMathOperations.
 * These tests are intentionally flaky to test flake detection systems.
 * 
 * - generateRandomOddNumber() test fails ~25% of the time
 * - generateRandomEvenNumber() test fails ~50% of the time
 * - generateRandomPrimeCandidate() test fails ~75% of the time
 */
@DisplayName("RandomMathOperations Tests (Intentionally Flaky)")
class RandomMathOperationsTest {

    private final RandomMathOperations randomOps = new RandomMathOperations();

    @RepeatedTest(value = 20, name = "Run {currentRepetition}/{totalRepetitions}")
    @DisplayName("Random odd number should be odd (Flaky: fails ~25% of the time)")
    void testGenerateRandomOddNumber() {
        int number = randomOps.generateRandomOddNumber();
        assertTrue(number >= 1 && number <= 100, "Number should be between 1 and 100");
        assertTrue(number % 2 == 1, "Number should be odd, but got: " + number);
    }

    @RepeatedTest(value = 20, name = "Run {currentRepetition}/{totalRepetitions}")
    @DisplayName("Random even number should be even (Flaky: fails ~50% of the time)")
    void testGenerateRandomEvenNumber() {
        int number = randomOps.generateRandomEvenNumber();
        assertTrue(number >= 0 && number <= 100, "Number should be between 0 and 100");
        assertTrue(number % 2 == 0, "Number should be even, but got: " + number);
    }

    @RepeatedTest(value = 20, name = "Run {currentRepetition}/{totalRepetitions}")
    @DisplayName("Random prime candidate should be prime (Flaky: fails ~75% of the time)")
    void testGenerateRandomPrimeCandidate() {
        int number = randomOps.generateRandomPrimeCandidate();
        assertTrue(number >= 2 && number <= 200, "Number should be between 2 and 200");
        assertTrue(RandomMathOperations.isPrime(number), "Number should be prime, but got: " + number);
    }

    @Test
    @DisplayName("Helper method isPrime should correctly identify prime numbers")
    void testIsPrime() {
        // Test prime numbers
        assertTrue(RandomMathOperations.isPrime(2));
        assertTrue(RandomMathOperations.isPrime(3));
        assertTrue(RandomMathOperations.isPrime(5));
        assertTrue(RandomMathOperations.isPrime(7));
        assertTrue(RandomMathOperations.isPrime(11));
        assertTrue(RandomMathOperations.isPrime(97));

        // Test non-prime numbers
        assertFalse(RandomMathOperations.isPrime(1));
        assertFalse(RandomMathOperations.isPrime(4));
        assertFalse(RandomMathOperations.isPrime(6));
        assertFalse(RandomMathOperations.isPrime(9));
        assertFalse(RandomMathOperations.isPrime(100));
    }
}

File: kotlin/src/test/kotlin/com/corntest/RandomMathOperationsTest.kt
package com.corntest

import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.test.assertEquals

class RandomMathOperationsTest {

    @Test
    fun testGenerateRandomOddNumber() {
        val rng = RandomMathOperations()
        repeat(20) { iteration ->
            val result = rng.generateRandomOddNumber()
            assertTrue(
                result % 2 != 0,
                "Iteration $iteration: Expected odd number but got $result"
            )
            assertTrue(
                result in 1..99,
                "Iteration $iteration: Expected value in 1..99 but got $result"
            )
        }
    }

    @Test
    fun testGenerateRandomEvenNumber() {
        val rng = RandomMathOperations()
        repeat(20) { iteration ->
            val result = rng.generateRandomEvenNumber()
            assertEquals(
                0, result % 2,
                "Iteration $iteration: Expected even number but got $result"
            )
            assertTrue(
                result in 0..100,
                "Iteration $iteration: Expected value in 0..100 but got $result"
            )
        }
    }

    @Test
    fun testGenerateRandomPrimeCandidate() {
        val rng = RandomMathOperations()
        repeat(20) { iteration ->
            val result = rng.generateRandomPrimeCandidate()
            assertTrue(
                RandomMathOperations.isPrime(result),
                "Iteration $iteration: Expected prime but got $result"
            )
            assertTrue(
                result in 2..97,
                "Iteration $iteration: Expected value in 2..97 but got $result"
            )
        }
    }

    @Test
    fun testIsPrimeWithKnownPrimes() {
        val primes = listOf(2, 3, 5, 7, 11, 13, 17, 19, 23, 29)
        for (p in primes) {
            assertTrue(RandomMathOperations.isPrime(p), "$p should be prime")
        }
    }

    @Test
    fun testIsPrimeWithNonPrimes() {
        val nonPrimes = listOf(0, 1, 4, 6, 8, 9, 10, 12, 15, 100)
        for (n in nonPrimes) {
            assertTrue(!RandomMathOperations.isPrime(n), "$n should not be prime")
        }
    }
}
